### PR TITLE
Support `eye` for sparse layouts (CSR and COO)

### LINKED ISSUE #46705: torch.eye not supported for sparse layout
1. Call torch.eye with layout = torch.sparse_coo

Below is a code sample:

import torch

x = torch.eye(4, layout=torch.sparse_coo, device=torch.device('cuda'))
print(x)

## Actual behavior
Traceback (most recent call last):
  File "sparse_tensor.py", line 46, in <module>
    x = torch.eye(4, layout=torch.sparse_coo, device=torch.device('cuda'))
RuntimeError: Could not run 'aten::eye.m_out' with arguments from the 'SparseCUDATensorId' backend. 'aten::eye.m_out' is only available for these backends: [CPUTensorId, CUDATensorId, VariableTensorId].

## Expected behavior
The Sparse Tensor should get created and printed properly as below:

tensor(indices=tensor([[0, 1, 2, 3],
                       [0, 1, 2, 3]]),
       values=tensor([1., 1., 1., 1.]),
       size=(4, 4), nnz=4, layout=torch.sparse_coo)

## Environment

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.8.0a0
[pip3] torchtext==0.8.0a0+7e267d2
[pip3] torchvision==0.8.0a0+be8192e
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] magma-cuda102             2.5.2                         1    pytorch
[conda] mkl                       2020.1                      217  
[conda] mkl-include               2020.1                      217  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.1.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.1           py38hbc911f0_0  
[conda] numpy-base                1.19.1           py38hfa32c7d_0  
[conda] torch                     1.8.0a0                   dev_0    <develop>
[conda] torchtext                 0.8.0a0+7e267d2          pypi_0    pypi
[conda] torchvision               0.8.0a0+be8192e          pypi_0    pypi

 - PyTorch Version: 1.8.0a0
 - OS: Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): python setup.py develop
 - Python version: 3.8
 - CUDA/cuDNN version: 10.2

### PR DESCRIPTION
**Pull request by krshrimali**
Fixes https://github.com/pytorch/pytorch/issues/46705. This PR adds support for `eye` for both Sparse CSR and COO layouts.

cc: @IvanYashchuk @pearu @cpuhrsch 

P.S: Thanks to @hlahkar for his work for Sparse COO layout here: https://github.com/pytorch/pytorch/pull/46709. :)

### CODE REVIEW

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -1409,6 +1409,18 @@ def test_exercise_detach(self, device, dtype):
             self.assertEqual(inp.crow_indices(), detached_inp.crow_indices())
             self.assertEqual(inp.col_indices(), detached_inp.col_indices())
 
+    @skipMeta
+    @dtypes(*get_all_dtypes(include_bfloat16=False, include_complex=False))
```
Comment (IvanYashchuk) [inline test/test_sparse_csr.py:1413]: Do tests pass for `include_bfloat16=True` and `include_complex=True`?

Comment (krshrimali) [reply test/test_sparse_csr.py:1413]: Thanks, @IvanYashchuk for asking this question. It only fails for `bfloat16`, I've revised the types in the latest comment. Also for Sparse COO, `eye` supports `complex64, complex128` but not `complex32`, so I've revised that as well.

Comment (IvanYashchuk) [reply test/test_sparse_csr.py]: What's the error message for `bfloat16`?

Comment (krshrimali) [reply test/test_sparse_csr.py]: From what observed, the reasons are 2:

1. `eye` not implemented for `BFloat16`
2. `to_dense()` doesn't work with `BFloat16` (for Sparse COO)

Following is the error that I'm getting:

```python
======================================================================
ERROR: test_eye_cpu_bfloat16 (__main__.TestSparseCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/pytorch/test/test_sparse.py", line 3412, in test_eye
    dense_output = torch.eye(n, device=device, dtype=dtype)
RuntimeError: "eye" not implemented for 'BFloat16'
```

Comment (IvanYashchuk) [reply test/test_sparse_csr.py]: It works on CUDA.
```py
In [1]: import torch

In [2]: torch.eye(3, dtype=torch.bfloat16)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-27c993df381a> in <module>
----> 1 torch.eye(3, dtype=torch.bfloat16)

RuntimeError: "eye" not implemented for 'BFloat16'

In [3]: torch.eye(3, dtype=torch.bfloat16, device='cuda')
Out[3]:
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]], device='cuda:0', dtype=torch.bfloat16)

In [4]: torch.eye(3, dtype=torch.bfloat16, device='cuda').to_sparse()
Out[4]:
tensor(indices=tensor([[0, 1, 2],
                       [0, 1, 2]]),
       values=tensor([1., 1., 1.]),
       device='cuda:0', size=(3, 3), nnz=3, dtype=torch.bfloat16,
       layout=torch.sparse_coo)

In [5]: torch.eye(3, dtype=torch.bfloat16, device='cuda').to_sparse().to_dense()
Out[5]:
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]], device='cuda:0', dtype=torch.bfloat16)
```

Comment (IvanYashchuk) [reply test/test_sparse_csr.py]: Enabling `eye` for bfloat16 on CPU shouldn't be much of an effort. Probably requires just changing the dispatch macro.

Comment (krshrimali) [reply test/test_sparse_csr.py]: Right, but then `to_dense()` will still fail for `BFloat16` types on CPU.

```python
ERROR: test_eye_cpu_bfloat16 (__main__.TestSparseCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 389, in instantiated_test
    raise rte
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 796, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/home/krshrimali/pytorch/test/test_sparse.py", line 3414, in test_eye
    self.assertEqual(sparse_coo_output.to_dense(), dense_output)
RuntimeError: "add_dense_sparse" not implemented for 'BFloat16'
```

Locally, in the `add_out_sparse_cpu` function in `SparseTensorMath.cpp` file, if I change the dispatch macro to add `BFloat16`, the test passes. I'm pushing these commits, if you can take a look - that will help!

Comment (krshrimali) [reply test/test_sparse_csr.py]: > Locally, in the `add_out_sparse_cpu` function in `SparseTensorMath.cpp` file, if I change the dispatch macro to add `BFloat16`, the test passes. I'm pushing these commits, if you can take a look - that will help!

I tried adding `Half` in the dispatch macro as well, and the test passes with `Half` as a dtype. Should we do that? Or was there a reason that it wasn't done before?

Comment (krshrimali) [reply test/test_sparse_csr.py]: This should be ready for review now :)

**Comment by krshrimali** [comment]
@IvanYashchuk @cpuhrsch - Can you please take a look at this PR? Will appreciate it if this can be merged after review.

**Comment by IvanYashchuk** [comment]
I expect https://github.com/pytorch/pytorch/pull/72443 to be merged first, then this PR needs to update the function signature.

**Pull request closed without merging**